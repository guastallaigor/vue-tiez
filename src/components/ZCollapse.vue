<template>
  <div
    class="z-collapse"
    v-if="copiedItems && copiedItems.length"
    :style="maxWidth ? `max-width:${maxWidth}` : ''"
  >
    <div
      class="collapse"
      v-for="(item, index) in copiedItems"
      :key="index"
      :class="{'dark': dark}"
      @click="changeHeight(item)"
    >
      <h3 role="button" :tabindex="index" :aria-expanded="item.height > 0">
        <span>{{ item[titleAttr] }}</span>
        <span>
          <svg width="15" height="15" fill="none" aria-label="open" viewBox="0 0 15 15" v-show="item.height === 0">
            <line x1="7.5" y1="2.18557e-08" x2="7.5" y2="15" :stroke="getStroke"></line>
            <line x1="15" y1="7.5" x2="-4.37114e-08" y2="7.5" :stroke="getStroke"></line>
          </svg>
          <svg width="15" height="15" fill="none" aria-label="close" viewBox="0 0 15 15" v-show="item.height > 0">
            <line x1="15" y1="7.5" x2="-4.37114e-08" y2="7.5" :stroke="getStroke"></line>
          </svg>
        </span>
      </h3>
      <div class="collapse-content" :style="`height:${item.height}px;`">
        <div>
          <p>{{ item[contentAttr] }}</p>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'z-collapse',
  props: {
    items: {
      type: Array,
      required: true
    },
    maxWidth: {
      type: String,
      default: '600px'
    },
    contentAttr: {
      type: String,
      default: 'content'
    },
    titleAttr: {
      type: String,
      default: 'title'
    },
    dark: {
      type: Boolean,
      default: false
    }
  },
  data: () => ({
    copiedItems: []
  }),
  computed: {
    getStroke () {
      return this.dark ? '#fff' : '#000'
    }
  },
  created () {
    this.setCopiedItems()
  },
  methods: {
    setCopiedItems () {
      this.copiedItems = this.items && this.items.length
        ? this.items.map(it => ({ ...it, height: 0 }))
        : []
    },
    changeHeight (item) {
      item.height = (item.height > 0) ? 0 : 120

      this.copiedItems.forEach(it => {
        if ((item[this.titleAttr] !== it[this.titleAttr]) &&
          (item[this.contentAttr] !== it[this.contentAttr])) {
          it.height = 0
        }
      })
    }
  }
}
</script>

<style lang="scss">
.z-collapse {
  border-top: 1px solid rgb(234, 234, 234);

  .collapse {
    width: 100%;
    text-align: left;
    border-bottom: 1px solid rgb(234, 234, 234);

    h3 {
      display: flex;
      align-items: center;
      justify-content: space-between;
      min-height: 75px;
      font-size: 20px;
      font-weight: 600;
      text-transform: none;
      color: rgb(0, 0, 0);
      cursor: pointer;
      outline-width: 0;
      margin: 0;

      > span:first-child {
        margin-right: 1em;
        line-height: 1.35em;
        padding: 15px 0;
      }
    }

    .collapse-content {
      overflow-y: hidden;
      transition: height 0.2s ease 0s;

      > div {
        padding-bottom: 20px;
        overflow-y: hidden;

        p {
          font-weight: 400;
          font-size: 14px;
          line-height: 24px;
          color: rgb(0, 0, 0);
        }
      }
    }

    &.dark {
      h3,
      div p {
        color: rgb(255, 255, 255);
      }
    }
  }
}
</style>
