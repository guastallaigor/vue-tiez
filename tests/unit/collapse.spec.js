import { shallowMount } from '@vue/test-utils'
import ZCollapse from '../../src/components/ZCollapse'

describe('When I create the ZCollapse component', () => {
  const createComponent = (propsData = {}, slot) => {
    return shallowMount(ZCollapse, {
      propsData: propsData,
      slots: {
        default: slot || 'ZCollapse'
      }
    })
  }

  const items = [
    {
      title: 'Question A',
      content: 'Lorem ipsum dolor sit amet'
    },
    {
      title: 'Question B',
      content: 'Duis aute irure dolor in reprehenderit'
    }
  ]

  it('should be a Vue instance', () => {
    const wrapper = createComponent({ items })
    expect(wrapper.find(ZCollapse).isVueInstance()).toBe(true)
  })

  it('should have a default z-collapse state, with title and content', () => {
    const wrapper = createComponent({ items })
    const collapse = wrapper.find('.z-collapse')
    expect(collapse.exists()).toBe(true)
    expect(collapse.classes().length).toBe(1)
    expect(collapse.attributes().style).toBe('max-width: 600px;')
    const collapseDiv = wrapper.findAll('.z-collapse > .collapse')
    expect(collapseDiv.at(0).exists()).toBe(true)
    expect(collapseDiv.at(1).exists()).toBe(true)
    expect(collapseDiv.length).toBe(2)
    const title = wrapper.findAll('.z-collapse > .collapse > h3 > span:first-child')
    expect(title.at(0).exists()).toBe(true)
    expect(title.at(1).exists()).toBe(true)
    expect(title.length).toBe(2)
    expect(title.at(0).text()).toBe('Question A')
    expect(title.at(1).text()).toBe('Question B')
    const content = wrapper.findAll('.z-collapse > .collapse > .collapse-content > div > p')
    expect(content.at(0).exists()).toBe(true)
    expect(content.at(1).exists()).toBe(true)
    expect(content.length).toBe(2)
    expect(content.at(0).text()).toBe('Lorem ipsum dolor sit amet')
    expect(content.at(1).text()).toBe('Duis aute irure dolor in reprehenderit')
    const collapseContent = wrapper.findAll('.z-collapse > .collapse > .collapse-content')
    expect(collapseContent.at(0).exists()).toBe(true)
    expect(collapseContent.at(1).exists()).toBe(true)
    expect(collapseContent.length).toBe(2)
    expect(collapseContent.at(0).attributes().style).toBe('height: 0px;')
    expect(collapseContent.at(1).attributes().style).toBe('height: 0px;')
    const spanSvg = wrapper.findAll('.z-collapse > .collapse > h3 > span:last-child > svg')
    expect(spanSvg.at(0).exists()).toBe(true)
    expect(spanSvg.at(1).exists()).toBe(true)
    expect(spanSvg.at(1).attributes().style).toBe('display: none;')
    expect(spanSvg.at(2).exists()).toBe(true)
    expect(spanSvg.at(3).exists()).toBe(true)
    expect(spanSvg.at(3).attributes().style).toBe('display: none;')
    const spanLine = wrapper.findAll('.z-collapse > .collapse > h3 > span:last-child > svg > line')
    expect(spanLine.at(0).exists()).toBe(true)
    expect(spanLine.at(0).attributes().stroke).toBe('#000')
    expect(spanLine.at(1).exists()).toBe(true)
    expect(spanLine.at(1).attributes().stroke).toBe('#000')
    expect(spanLine.at(2).exists()).toBe(true)
    expect(spanLine.at(2).attributes().stroke).toBe('#000')
    expect(spanLine.at(3).exists()).toBe(true)
    expect(spanLine.at(3).attributes().stroke).toBe('#000')
    expect(spanSvg.length).toBe(4)
  })

  it('should have a dark class', () => {
    const wrapper = createComponent({ items, dark: true })
    const spanLine = wrapper.findAll('.z-collapse > .collapse > h3 > span:last-child > svg > line')
    const collapseDiv = wrapper.find('.z-collapse > .collapse')
    expect(collapseDiv.exists()).toBe(true)
    expect(collapseDiv.classes().length).toBe(2)
    expect(collapseDiv.classes()).toEqual(['collapse', 'dark'])
    expect(spanLine.at(0).exists()).toBe(true)
    expect(spanLine.at(0).attributes().stroke).toBe('#fff')
    expect(spanLine.at(1).exists()).toBe(true)
    expect(spanLine.at(1).attributes().stroke).toBe('#fff')
    expect(spanLine.at(2).exists()).toBe(true)
    expect(spanLine.at(2).attributes().stroke).toBe('#fff')
    expect(spanLine.at(3).exists()).toBe(true)
    expect(spanLine.at(3).attributes().stroke).toBe('#fff')
    expect(spanLine.length).toBe(6)
  })

  it('should work with a different content attribute', () => {
    const items = [
      {
        title: 'Question A',
        value: 'Lorem ipsum dolor sit amet'
      },
      {
        title: 'Question B',
        value: 'Duis aute irure dolor in reprehenderit'
      }
    ]

    const wrapper = createComponent({ items, contentAttr: 'value' })
    const content = wrapper.findAll('.z-collapse > .collapse > .collapse-content > div > p')
    expect(content.at(0).exists()).toBe(true)
    expect(content.at(1).exists()).toBe(true)
    expect(content.length).toBe(2)
    expect(content.at(0).text()).toBe('Lorem ipsum dolor sit amet')
    expect(content.at(1).text()).toBe('Duis aute irure dolor in reprehenderit')
  })

  it('should work with a different title attribute', () => {
    const items = [
      {
        value: 'Question A',
        content: 'Lorem ipsum dolor sit amet'
      },
      {
        value: 'Question B',
        content: 'Duis aute irure dolor in reprehenderit'
      }
    ]

    const wrapper = createComponent({ items, titleAttr: 'value' })
    const title = wrapper.findAll('.z-collapse > .collapse > h3 > span:first-child')
    expect(title.at(0).exists()).toBe(true)
    expect(title.at(1).exists()).toBe(true)
    expect(title.length).toBe(2)
    expect(title.at(0).text()).toBe('Question A')
    expect(title.at(1).text()).toBe('Question B')
  })

  it('should have a different max-width', () => {
    const wrapper = createComponent({ items, maxWidth: '10em' })
    const collapse = wrapper.find('.z-collapse')
    expect(collapse.exists()).toBe(true)
    expect(collapse.classes().length).toBe(1)
    expect(collapse.attributes().style).toBe('max-width: 10em;')
  })

  it('should be able to expand and collapse', () => {
    const wrapper = createComponent({ items })
    wrapper.setMethods({ changeHeight: jest.fn() })
    const collapse = wrapper.find('.z-collapse')
    expect(collapse.exists()).toBe(true)
    const collapseDiv = wrapper.find('.z-collapse > .collapse:first-child')
    collapseDiv.trigger('click')
    expect(wrapper.vm.changeHeight).toHaveBeenCalledTimes(1)
    // WIP - NEEDS IMPROVEMENT
    // const collapseContent = wrapper.findAll('.z-collapse > .collapse > .collapse-content')
    // expect(collapseContent.at(0).exists()).toBe(true)
    // expect(collapseContent.at(0).attributes().style).toBe('height: 120px;')
  })

  it('should match snapshot', () => {
    const wrapper = createComponent({
      items,
      maxWidth: '600px',
      contentAttr: 'content',
      titleAttr: 'title',
      dark: false
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
