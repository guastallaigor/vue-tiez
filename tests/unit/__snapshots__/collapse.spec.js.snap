// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the ZCollapse component should match snapshot 1`] = `
<div class="z-collapse" style="max-width: 600px;">
  <div class="collapse">
    <h3 role="button" tabindex="0"><span>Question A</span> <span><svg width="15" height="15" fill="none" aria-label="open" viewBox="0 0 15 15"><line x1="7.5" y1="2.18557e-08" x2="7.5" y2="15" stroke="#000"></line> <line x1="15" y1="7.5" x2="-4.37114e-08" y2="7.5" stroke="#000"></line></svg> <svg width="15" height="15" fill="none" aria-label="close" viewBox="0 0 15 15" style="display: none;"><line x1="15" y1="7.5" x2="-4.37114e-08" y2="7.5" stroke="#000"></line></svg></span></h3>
    <div class="collapse-content" style="height: 0px;">
      <div>
        <p>Lorem ipsum dolor sit amet</p>
      </div>
    </div>
  </div>
  <div class="collapse">
    <h3 role="button" tabindex="1"><span>Question B</span> <span><svg width="15" height="15" fill="none" aria-label="open" viewBox="0 0 15 15"><line x1="7.5" y1="2.18557e-08" x2="7.5" y2="15" stroke="#000"></line> <line x1="15" y1="7.5" x2="-4.37114e-08" y2="7.5" stroke="#000"></line></svg> <svg width="15" height="15" fill="none" aria-label="close" viewBox="0 0 15 15" style="display: none;"><line x1="15" y1="7.5" x2="-4.37114e-08" y2="7.5" stroke="#000"></line></svg></span></h3>
    <div class="collapse-content" style="height: 0px;">
      <div>
        <p>Duis aute irure dolor in reprehenderit</p>
      </div>
    </div>
  </div>
</div>
`;
